### PR TITLE
Vulkan Renderer

### PR DESCRIPTION
**Pull request by danginsburg**
## Description
This pull request adds an implementation of a Vulkan Render backend to SDL.  I have so far tested this primarily on Windows, but also smoke tested on Linux and macOS (MoltenVK).  I have not tried it yet on Android, but it should be usable there as well (sans any bugs I missed).  This began as a port of the SDL Direct3D12 Renderer, which is the closest thing to Vulkan as existed in the SDL codebase. The shaders are more or less identical (with the only differences being in descriptor bindings vs root descriptors).  The shaders are built using the HLSL frontend of glslang.

Everything in the code is pure Vulkan 1.0 (no extensions), with the exception of HDR support which requires the Vulkan instance extension `VK_EXT_swapchain_colorspace`.  The code could have been simplified considerably if I used dynamic rendering, push descriptors, extended dynamic state, and other modern Vulkan-isms, but I felt it was more important to make the code as vanilla Vulkan as possible so that it would run on any Vulkan implementation.

The main differences with the Direct3D12 renderer are:
* Having to manage renderpasses for performing clears.  There is likely some optimization that would still remain for more efficient use of TBDR hardware where there might be some unnecessary load/stores, but it does attempt to do clears using renderpasses.
* Constant buffer data couldn't be directly updated in the command buffer since I didn't want to rely on push descriptors, so there is a persistently mapped buffer with increasing offset per swapchain image where CB data gets written.
* Many more resources are dependent on the swapchain resizing due to i.e. Vulkan requiring the VkFramebuffer to reference the VkImageView of the swapchain, so there is a bit more code around handling that than was necessary in D3D12.
* For NV12/NV21 textures, rather than there being plane data in the texture itself, the UV data is placed in a separate `VkImage`/`VkImageView`.

## What I've Tested
I've verified that `testcolorspace` works with both sRGB and HDR linear.  I've tested `testoverlay` works with the various YUV/NV12/NV21 formats.  I've tested `testsprite`.  I've checked that window resizing and swapchain out-of-date handling when minimizing are working.  I've run through `testautomation` with the render tests.  I also have run several of the tests with Vulkan validation and synchronization validation.  Surely I will have missed some things, but I think it's in a good state to be merged and build out from here.

### CODE REVIEW

File: include/SDL3/SDL_hints.h [hunk 1]
```diff
@@ -1737,7 +1737,19 @@ extern "C" {
...
+ *  This variable can be set to the following values:
+ *    "0"       - Disable Validation Layer use
+ *    "1"       - Enable Validation Layer use
+ *
+ *  By default, SDL does not use Vulkan Validation Layers.
+ */
+#define SDL_HINT_RENDER_VULKAN_DEBUG    "SDL_RENDER_VULKAN_DEBUG"
+
+/**
+ *  A variable controlling whether the Direct3D device is initialized for thread-safe operations.
```
Comment (madebr) [inline include/SDL3/SDL_hints.h:1752]:
```suggestion
 * A variable specifying which render driver to use.
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -335,6 +335,7 @@ dep_option(SDL_RENDER_D3D          "Enable the Direct3D render driver" ON "SDL_R
 dep_option(SDL_RENDER_METAL        "Enable the Metal render driver" ON "SDL_RENDER;${APPLE}" OFF)
 dep_option(SDL_VIVANTE             "Use Vivante EGL video driver" ON "${UNIX_SYS};SDL_CPU_ARM32" OFF)
 dep_option(SDL_VULKAN              "Enable Vulkan support" ON "SDL_VIDEO;ANDROID OR APPLE OR LINUX OR WINDOWS" OFF)
+dep_option(SDL_RENDER_VULKAN       "Enable the Vulkan render driver" ON "SDL_VIDEO;ANDROID OR APPLE OR LINUX OR WINDOWS" OFF)
```
Comment (madebr) [inline CMakeLists.txt:338]:
```suggestion
dep_option(SDL_RENDER_VULKAN       "Enable the Vulkan render driver" ON "SDL_RENDER;SDL_VULKAN;ANDROID OR APPLE OR LINUX OR WINDOWS" OFF)
```

File: include/SDL3/SDL_hints.h (hunk 1 above)
Comment (danginsburg) [reply include/SDL3/SDL_hints.h:1752]: Oh, yes, this was a merge error, thanks.

**Comment by sezero** [comment]
Is the priority of `VULKAN_RenderDriver` in `render_drivers[]` good?

**Comment by slouken** [comment]
Yep, we'll leave it last for now, and we can bump up the priority if we find it's valuable to do so on some platforms.

**Comment by sezero** [comment]
We are using vulkan library entries as our vulkan function pointer names in SDL_render_vulkan.c without renaming, e.g. `vkMapMemory`. I hope it doesn't bite us in the future (like #2706 ...)

**Comment by slouken** [comment]
Yeah, I've been thinking about moving them into the renderer data, like I did with testffmpeg_vulkan.c. What do you think?

**Comment by sezero** [comment]
> Yeah, I've been thinking about moving them into the renderer data, like I did with testffmpeg_vulkan.c. What do you think?

Yes, I guess. @danginsburg to decide, better?

**Comment by danginsburg** [comment]
> Yes, I guess. @danginsburg to decide, better?

Sounds like a good idea, yeah.  I copy-pasted the function binding piece from testvulkan.c.  Normally I've used volk (https://github.com/zeux/volk) but assumed we didn't want the external dependency in SDL.

**Pull request merged**